### PR TITLE
GLGUI: chat - added comments and minor runtime optimizations

### PR DESCRIPTION
**Pull request by 0-8-15**
Additionally I had to re-format the source according to widespread
conventions for readbility and used keyword like `when` and `unless`
to make the source more compact and easier to digest.

Sorry for the latter.

### CODE REVIEW

**Comment by mgorges** [comment]
I have to disagree - there is not a single place in the entire codebase, where `unless` or `when` are used, as such this might confuse instead of help. You said not to make personal style related changed in one of your last pull request, but to me this clearly is, as this is your preference, not what the project has as a whole.

**Comment by mgorges** [comment]
To make sure this is documented. The expression equivalence for `when` is:

``` 
(when <test>
  <expression1>
  <expression2>
)
```
is equivalent to
```
(if <test>
  (begin
    <expression1>
    <expression2> 
  )
)
```

**Comment by mgorges** [comment]
The other case for `unless` includes an additional `not`, so:
```
(unless <test>
  <expression1>
  <expression2>
)
```
is equivalent to 
```
(if (not <test>)
  (begin
    <expression1>
    <expression2> 
  )
) 
```

File: modules/ln_glgui/chat.scm [hunk 1]
```diff
@@ -204,54 +204,59 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
...
-))
+         (inside (and (fx> mx-fixed x) (fx< mx-fixed (fx+ x w)) (fx> my-fixed y) (fx< my-fixed (fx+ y h)))))
+    (cond
+     ((and (fx= type EVENT_MOTION) old) ;; drag
+      (if (> (abs (- my fsty)) 5)
+          (let* ((calculated (+ val (* (- valmax valmin) (/ (flo (* 1. (- oldy my))) (flo h))))) ;; Why these `flo`?
+                 (clipped (max valmin (min valmax calculated))))
+            (glgui-widget-set! g wgt 'offset clipped)
+            (glgui-widget-set! g wgt 'oldy my-fixed)
+            (unless drag (glgui-widget-set! g wgt 'drag #t)))))
```
Comment (mgorges) [inline modules/ln_glgui/chat.scm:234]: This piece is new, it tests if `drag` is already set.

File: modules/ln_glgui/chat.scm [hunk 2]
```diff
@@ -204,54 +204,59 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
...
+     ((and inside (fx= type EVENT_BUTTON1DOWN)) ;; touch down
+      (glgui-widget-set! g wgt 'oldy my-fixed)
+      (glgui-widget-set! g wgt 'fsty my-fixed)
+      (glgui-widget-set! g wgt 'old #t)
+      (glgui-widget-set! g wgt 'drag #f))
+     ((and old (fx= type EVENT_BUTTON1UP)) ;; touch release
+      (cond
+       ((not inside) ;; cancel drag
+        (if old (glgui-widget-set! g wgt 'old #f))
+        (if drag (glgui-widget-set! g wgt 'drag #f)))
```
Comment (mgorges) [inline modules/ln_glgui/chat.scm:244]: this piece is new too.

File: modules/ln_glgui/chat.scm [hunk 3]
```diff
@@ -204,54 +204,59 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
...
+       (else
+        (glgui-widget-set! g wgt 'offset (round val)) ;; snap to place?
+        (if (and (not drag) inside cb)
+            (let ((cur (round (+ val (/ (- (+ y h) (+ my space)) dh) -.5))))
+              (when (and (>= cur 0) (>= cur val) (< cur n))
+                (glgui-widget-set! g wgt 'current cur)
+                (if (procedure? cb) (cb g wgt type mx my)))))
+        (if old (glgui-widget-set! g wgt 'old #f))
```
Comment (mgorges) [inline modules/ln_glgui/chat.scm:252]: new test added

File: modules/ln_glgui/chat.scm [hunk 4]
```diff
@@ -204,54 +204,59 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
...
+        (if drag (glgui-widget-set! g wgt 'drag #f)))))
+     (else ;; neither button1 nor motion
+      (unless inside
```
Comment (mgorges) [inline modules/ln_glgui/chat.scm:255]: this piece is also new

File: modules/ln_glgui/chat.scm [hunk 5]
```diff
@@ -204,54 +204,59 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 ))
 
 (define (glgui:chat-input g wgt type mx my)
-  (let* ((x (fix (glgui-widget-get-dyn g wgt 'x)))
+  (let* ((mx-fixed (fix mx)) ;; Q: Can `mx` or `my` ever not be fixnums?
```
Comment (mgorges) [inline modules/ln_glgui/chat.scm:207]: I don't know, but we use `fx` math later, which segfaults if it is not without pointing to where the error is, which is extremely annoying to debug.

**Comment by 0-8-15** [comment]
Am Sat, 26 Sep 2020 13:02:00 -0700
schrieb Matthias Görges <notifications@github.com>:

> I have to disagree - there is not a single place in the entire
> codebase, where `unless` or `when` are used, as such this might
> confuse instead of help. You said not to make personal style related
> changed in one of your last pull request, but to me this clearly is,
> as this is your preference, not what the project has as a whole.

Sorry for that.  --  As I wrote in the commit message.

Initially I just wanted to understand the code (since I was hunting for
the bug I made elsewhere).

Then (with the background of ~30 years of Scheme experience) I felt that
the unusual formatting was the part holding me back.  There is a rather
long tradition to use automatic code formatting among most Schemers.
Essentially the editors do this for you.  A few more rules are usually
recommended, e.g., multiple let-bindings on a line are OK if the column
76 is not crossed, but for let* and letrec this is never done; always
have a space between adjacent closing and opening parenthesis (since
some reader extensions like SRFI-110 or gambits 'six' might convert
these things); nothing behind `begin` if the block spans multiple lines.

Additionally working with `git` (which I don't really like for a
version control system, but it won the world) taught me the habit to
let the editor automatically remove trailing white space.

Then I ran into the repeated use of `length` for the same input.
`length` is O(n) and thus usually avoided for performance.  As we are
in the many-time-per-second part of the code I just removed that one.
Same for the repeated use of `fix` on the same input.  As I now was in
the optimized-a-few-thing-here-mode I added those "don't modify a hash
table for no reason" checks.  For consistency I added another case in
the else branch.

At that point I was convinced that I should now understand the code and
can't find a bug there.  Hence I looked an found it elsewhere.

What to do with the change now?  Throw away the improvements just
because they contain what I would call a code cleanup because that
would include personal preferences?  Or propose it as a PR and have you
guy decide?

I opted for the latter.

**Pull request merged**